data-prepper-api/src/main/java/org/opensearch/dataprepper/model/event/JacksonEvent.java

```diff
@@ -150,31 +150,33 @@ public JsonNode getJsonNode() {
         return jsonNode;
     }
 
-    void normalizeKeys(Map<String, Object> map) {
-        Map<String, Object> toAdd = new HashMap<>();
+    private Map<String, Object> normalizeKeys(final Map<String, Object> map) {
+        final Map<String, Object> replacementMap = new HashMap<>();
         Iterator<Map.Entry<String, Object>> iterator = map.entrySet().iterator();
         while (iterator.hasNext()) {
             Map.Entry<String, Object> entry = iterator.next();
             String key = entry.getKey();
             Object value = entry.getValue();
             final String newKey = JacksonEventKey.replaceInvalidCharacters(key);
             if (value instanceof Map) {
-                normalizeKeys((Map<String, Object>)value);
+                value = normalizeKeys((Map<String, Object>)value);
             }
             if (!newKey.equals(key)) {
-                toAdd.put(newKey, value);
-                iterator.remove();
+                replacementMap.put(newKey, value);
+            } else {
+                replacementMap.put(key, value);
```

Comment:
Couldn't you just use the `newKey` in both cases now?

Reply:
Good call. I just pushed a new revision with this.